share/p2p/shrex/server.go

```diff
@@ -0,0 +1,216 @@
+package shrex
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/ipfs/go-blockservice"
+	"github.com/ipfs/go-cid"
+	logging "github.com/ipfs/go-log/v2"
+	"github.com/libp2p/go-libp2p-core/host"
+	"github.com/libp2p/go-libp2p-core/network"
+	"github.com/minio/sha256-simd"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/celestiaorg/celestia-node/share"
+	"github.com/celestiaorg/celestia-node/share/eds"
+	"github.com/celestiaorg/celestia-node/share/ipld"
+	share_p2p_v1 "github.com/celestiaorg/celestia-node/share/p2p/v1/pb"
+	"github.com/celestiaorg/go-libp2p-messenger/serde"
+	"github.com/celestiaorg/nmt"
+	"github.com/celestiaorg/nmt/namespace"
+)
+
+const ndProtocolID = "/shrex/nd/0.0.1"
+
+var log = logging.Logger("shrex/nd")
+
+const (
+	writeTimeout = time.Second * 10
+	readTimeout  = time.Second * 10
+	serveTimeout = time.Second * 10
+)
+
+type Server struct {
+	store        *eds.Store
+	host         host.Host
+	writeTimeout time.Duration
+	readTimeout  time.Duration
+	serveTimeout time.Duration
```

Comment:
We can use the params pattern like in #1431 ? But non-blocking, can be done in a followup

Reply:
Will do a followup.